### ISSUE TITLE
[Feature Request] Return Attention Weights for `scaled_dot_product_attention`

### ISSUE THREAD

**Issue by aadyotb** [pytorch/pytorch#119811]
### 🚀 The feature, motivation and pitch

For many interpretability applications, it's quite useful to analyze the attention weights (`softmax(Q @ K.T)`) used to compute the output of an attention operation. However, these aren't currently returned by `scaled_dot_product_attention` (though they are returned by the flash attention library). It would be great to add an argument that optionally returns the attention weights.

cc @jbschlosser @bhosmer @cpuhrsch @erichan1 @drisspg @mikaylagawarecki

**Comment by davidbuterez** [comment]
I also think that this is an important feature and would love to see it implemented.

**Comment by drisspg** [comment]
Sdpa calls into one of a number of backends that perform fused attention - the algorithm defined in the flash attention paper. Part of the performance gains for these fused implementations is not having to read and write the intermediate attention scores to global memory. If you need the attention scores I would recommend writing your own version of sdpa in python - the docs for the function should provide a good basis for doing this

**Comment by Lifedecoder** [comment]
implemention could not reproduce results from F.scale_dot_product_attention, are there any other ways to output attention weights?

**Comment by MikeTkachuk** [comment]
You can always pass identity tensor instead of the actual values to get attention tensor

**Comment by Tomorrowdawn** [comment]
> [MikeTkachuk](/MikeTkachuk)

But the kv cache is growing and we will get an attention weight matrix of shape [L, S](query length, kv length). SDPA requires a value input of shape [N, ... , S, Ev], That means, only if S and Ev happens to be the same, otherwise you can't pass an identity tensor(which is of shape [S, S]).

**Comment by MikeTkachuk** [comment]
The goal is to inspect the attention matrix if I understood correctly.
You can supply values eye(S) of shape [..., S, S] so that the output was att [..., L, S] @ values [..., S, S], which is identity operation, so you get raw attentions. Correct me if I'm wrong.

**Comment by Tomorrowdawn** [comment]
Oh.... You are right. I just realized that the dimensions of value and key can be different. That's a good workaround, thank you. 

Additionally, here's a small suggestion. Users of SDPA typically care about speed, and this approach would result in two calls to SDPA kernel and the need to prepare an identity matrix. I believe it would be better to support this directly through the API interface.

**Comment by drisspg** [comment]
@Tomorrowdawn 
>sdpa calls into one of a number of backends that perform fused attention - the algorithm defined in the flash attention paper. Part of the performance gains for these fused implementations is not having to read and write the intermediate attention scores to global memory. If you need the attention scores I would recommend writing your own version of sdpa in python - the docs for the function should provide a good basis for doing this
